contracts/contracts/harvest/Harvester.sol

```diff
@@ -0,0 +1,376 @@
+
+        IERC20 token = IERC20(_tokenAddress);
+
+        // if changing token swap provider cancel existing allowance
+        if (
+            oldUniswapAddress != address(0) &&
+            oldUniswapAddress != _uniswapV2CompatibleAddr
+        ) {
+            token.safeApprove(oldUniswapAddress, 0);
+        }
+
+        // Give Uniswap infinite approval when needed
+        if (_uniswapV2CompatibleAddr != address(0)) {
+            if (
+                token.allowance(address(this), _uniswapV2CompatibleAddr) <
+                type(uint256).max / 2
+            ) {
+                token.safeApprove(_uniswapV2CompatibleAddr, 0);
+                token.safeApprove(_uniswapV2CompatibleAddr, type(uint256).max);
+            }
+        }
+
+        emit RewardTokenConfigUpdated(
+            _tokenAddress,
+            _allowedSlippageBps,
+            _harvestRewardBps,
+            _uniswapV2CompatibleAddr,
+            _liquidationLimit
+        );
+    }
+
+    /***************************************
+                    Rewards
+    ****************************************/
+
+    /**
+     * @dev Transfer token to governor. Intended for recovering tokens stuck in
+     *      contract, i.e. mistaken sends.
+     * @param _asset Address for the asset
+     * @param _amount Amount of the asset to transfer
+     */
+    function transferToken(address _asset, uint256 _amount)
+        external
+        onlyGovernor
+    {
+        IERC20(_asset).safeTransfer(governor(), _amount);
+    }
+
+    /**
+     * @dev Collect reward tokens from all strategies
+     */
+    function harvest() external onlyGovernor nonReentrant {
+        _harvest();
+    }
+
+    /**
+     * @dev Swap all supported swap tokens for stablecoins via Uniswap.
+     */
+    function swap() external onlyGovernor nonReentrant {
+        _swap();
+    }
+
+    /*
+     * @dev Collect reward tokens from all strategies and swap for supported
+     *      stablecoin via Uniswap
+     */
+    function harvestAndSwap() external nonReentrant {
+        // TODO: add protection so that harvestAndSwap isn't called twice too closely together
+        _harvest();
+        _swap();
+        if (msg.sender != vaultAddress) {
+            IVault(vaultAddress).rebase();
+        }
+    }
+
+    /**
+     * @dev Collect reward tokens from all strategies
+     */
+    function _harvest() internal {
+        address[] memory allStrategies = IVault(vaultAddress)
+            .getAllStrategies();
+        for (uint256 i = 0; i < allStrategies.length; i++) {
+            _harvest(allStrategies[i]);
+        }
+    }
+
+    /**
+     * @dev Collect reward tokens for a specific strategy. Called from the vault.
+     * @param _strategyAddr Address of the strategy to collect rewards from
+     */
+    function harvest(address _strategyAddr) external onlyGovernor nonReentrant {
+        // TODO: add protection so that harvestAndSwap isn't called twice too closely together
```

Comment:
I think for simplicity, we can leave this too closely check out of the contract. It's really only useful for callers, and I think there are other ways for them to handle it.

Reply:
OK removing for now. I guess we can address this issue if we hear a lot of complaints from the ones using it. And not leave out premature optimisation 👍